CLI: module for new CCM CLI (and ccm script) (Preview)

stdweird opened the pull request:
A new module for extensible and unittested commandline interface, incl `ccm` script that uses it and similar extensible and unittested tabcompletion.
Based on #62 and #64

ned21 commented on src/main/bin/ccm:
```diff
@@ -0,0 +1,60 @@
+
+=head1 OPTIONS
+
+For the list op options, check the C<EDG::WP4::CCM::CLI> and
```
Redundant "op" before options?

ned21 replied:
ah - I guess you mean *of* not op?

ned21 commented on src/main/conf/quattor-ccm:
```diff
@@ -0,0 +1,240 @@
+{
+    local dir bin fn
+    fn="$1"
+    # use the "slow" sed here
```
What's "slow" about it? Or is this a note it should be replaced by something faster? e.g. ${PATH//:/ }

stdweird replied:
this function looks for the abspath of e.g. `sed`, using `sed`, so this `sed` is slow because of the PATH lookup. i'll make the comment a bit clearer

ned21 replied:
OK but - and excuse my ignorance of bash here it's not a language I use - why use sed when bash can do that natively? Or is my version of bash more recent than the average user we want to support?

ned21 commented on src/main/perl/Options.pm:
```diff
@@ -246,6 +259,39 @@ sub action_showcids
+    my ($self, $newactions) = @_;
+
+    while (my ($action, $help) = each %$newactions) {
+        my $method = $self->can("action_$action");
```
This variable seems a bit redundant? Is there a reason not to avoid the extra variable and call:
``` if($self->can("action_$action")) {```

stdweird commented:
retest this please

stdweird commented:
@ned21 wrt the bash buitins, i'm quite sure previous EL5 releases shipped bash that didn't have such support. i just checked on a 5.11 box, it does have it, but TUV could have backported this. the code you commented on is only run once during the souce-ing of the tabcompletion script, so it's not slowing the actual tabcompletion.

ned21 commented:
OK, makes sense - I make no claims to be competent at bash-ing :-) Any chance of a fix to the docs - there's a typo when you said op instead of of and the C<..> should be replaced by L<...> to link man pages together. Otherwise I think this is good to go.

stdweird commented:
@ned21 remarks fixed. could you also make a comment in https://github.com/quattor/CCM/pull/64 to indicate that that one is also fine (as this PR is based on that one).

ned21 commented:
I merged #64. Does anyone want to try out this new CLI before we merge it?

stdweird commented:
@ned21 as long as https://github.com/quattor/CAF/pull/98 is not resolved in some way, the output will be mixed with other output.
```
[root@host ~]# source /etc/bash_completion.d/quattor-ccm 
[root@host ~]# ccm --
--base_url          --cfgfile           --debug             --keep_old          --preprocessor      --retrieve_retries  --trust
--cache_root        --cid               --force             --key_file          --profile           --retrieve_wait     --world_readable
--ca_dir            --component         --format            --lock_retries      --profile_failover  --show              
--ca_file           --context           --get_timeout       --lock_wait         --profpath          --showcids          
--cert_file         --dbformat          --json_typed        --metaconfig        --purge_time        --tabcompletion     
[root@host ~]# ccm --for
--force   --format  
[root@host ~]# ccm --for
--force   --format  
[root@host ~]# ccm --format 
json           pan            pancxml        tabcompletion  yaml           
[root@host ~]# ccm --format json --profpath /system/
/system/aii/           /system/cluster/       /system/kernel/        /system/network/       /system/rootmail       /system/state
/system/blockdevices/  /system/filesystems/   /system/monitoring/    /system/quattorid      /system/siterelease    
[root@host ~]# ccm --format json --profpath /system/network/ --show
updated /var/lib/ccm/profile.9/ccm-active-profile.9-6562.writetest
updated /var/lib/ccm/profile.9/ccm-active-profile.9-6562
{"default_gateway":"1.2.3.4","domainname":"ugent.be","hostname":"host","interfaces":{"eth0":{"broadcast":"1.2.3.255","driver":"e1000","ip":"1.2.3.10","netmask":"255.255.255.0"}},"nameserver":["1.2.3.243"],"nozeroconf":true,"set_hwaddr":true}
```
